buckets/local/bucket.go

```diff
@@ -447,3 +448,107 @@ func addFile(dag ipld.DAGService, layout options.Layout, prefix cid.Prefix, r io
 	}
 	return n, err
 }
+
+// Diff returns a set of changes that transform node 'a' into node 'b'.
+// Modified from https://github.com/ipfs/go-merkledag/blob/master/dagutils/diff.go#L104
+// It only traverses links in the following cases:
+// 1. two node's links number are greater than 0.
+// 2. both of two nodes are ProtoNode.
+// 3. neither of the nodes is a file node, which contains only unnamed raw blocks
+// Otherwise, it compares the cid and emits a Mod change object.
+func Diff(ctx context.Context, ds ipld.DAGService, a, b ipld.Node) ([]*du.Change, error) {
+	// Base case where both nodes are leaves, just compare
+	// their CIDs.
+	if len(a.Links()) == 0 && len(b.Links()) == 0 {
+		return getChange(a, b)
+	}
+
+	var out []*du.Change
+	cleanA, okA := a.Copy().(*md.ProtoNode)
+	cleanB, okB := b.Copy().(*md.ProtoNode)
+	if !okA || !okB {
+		return getChange(a, b)
+	}
+
+	if isFileNode(cleanA) || isFileNode(cleanB) {
```

Comment:
If either is a file node, just mark as a change.

Reply:
So if we're here, that's because the link Cids of these files were different(some step before line 487 was `false`). Correct?

Reply:
Yep yep!